packages/editor/src/components/post-slug/index.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * WordPress dependencies
+ */
+import { withDispatch, withSelect } from '@wordpress/data';
+import { Component } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { withInstanceId, compose } from '@wordpress/compose';
+import { safeDecodeURIComponent } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import PostSlugCheck from './check';
+import { cleanForSlug } from '../../utils/url';
+
+export class PostSlug extends Component {
+	constructor( { postSlug, postTitle, postID } ) {
+		super( ...arguments );
+
+		this.state = {
+			editedSlug: safeDecodeURIComponent( postSlug ) || cleanForSlug( postTitle ) || postID,
```

Comment:
We shouldn't mirror state locally when this is already in the store.

Reply:
The reason that I added an internal `editedSlug` is that it is impossible to reliably edit the slug directly without an intermediate step.

* We need to call `cleanForSlug()` to make sure the value updated in the post is in valid slug format.
* However, if you call `cleanForSlug()` during editing (`onChange`), it will cause unexpected behavior for the user, mostly because trailing dashes are not allowed. So if you want to write e.g. `single-post` and are at `single-`, it will only show as `single`. Unless you copy/paste the whole thing or write `singlepost` and then add the dash (both of which is cumbersome), it is impossible to "naturally" write a dash.

Hence the intermediate step, and that's why the slug is only actually validated and updated `onBlur`.

Reply:
That makes sense :smile: